modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java

```diff
@@ -316,46 +323,33 @@ private void addFeatureInternal(SimpleFeature feature) {
 
         throw new IOException("Type name " + typeName + " not found");
     }
+    
+    
 
     /**
      * List of available types provided by this DataStore.
      *
-     * @return Array of type names
+     * @return List of type names
      *
-     * @see org.geotools.data.AbstractDataStore#getFeatureTypes()
+     * @see org.geotools.data.ContentDataStore#getFeatureTypes()
      */
-    public String[] getTypeNames() {
+    protected List<Name> createTypeNames() {
         synchronized (memory) {
-            String[] types = new String[schema.size()];
-            int index = 0;
-
-            for (Iterator<String> i = schema.keySet().iterator(); i.hasNext(); index++) {
-                types[index] = i.next();
+            List<Name> typeNames = new ArrayList<Name>();
+            
+            for (Iterator<String> i = schema.keySet().iterator(); i.hasNext();) {
+                typeNames.add( new NameImpl(namespaceURI, i.next()));
             }
-
-            return types;
+            return typeNames;
         }
     }
-
-    /**
-     * SimpleFeatureType access by <code>typeName</code>.
-     *
-     * @param typeName
-     *
-     * @return SimpleFeatureType for <code>typeName</code>
-     *
-     * @throws IOException
-     * @throws SchemaNotFoundException DOCUMENT ME!
-     *
-     * @see org.geotools.data.AbstractDataStore#getSchema(java.lang.String)
-     */
-    public SimpleFeatureType getSchema(String typeName) throws IOException {
-        synchronized (memory) {
-            if (schema.containsKey(typeName)) {
-                return schema.get(typeName);
-            }
-                throw new SchemaNotFoundException(typeName);
-        }
+    
+    protected ContentFeatureSource createFeatureSource(ContentEntry entry) {
+        return createFeatureSource(entry, Query.ALL);
+    }
+    
+    protected ContentFeatureSource createFeatureSource(ContentEntry entry, Query query) {
+        return new MemoryFeatureStore(entry, query);
```

Comment:
Memory is kind of wriable - so I am okay with FeatureStore all the time :)

If you really wanted you could make a constructor flag for read-only. We have DataUtilities.store( featureCollection ) class that can get you a read-only "memory" datastore for safely staging data.